src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/PreinitializationManager.cs

```diff
@@ -130,18 +124,33 @@ public void LogStatistics(Logger logger)
 
         public TypePreinit.PreinitializationInfo GetPreinitializationInfo(MetadataType type)
         {
-            return _preinitHashTable.GetOrCreateValue(type);
+            TypePreinit.PreinitializationInfo info = _preinitHashTable.GetOrCreateValue(type);
+#if DEBUG
+            TypeDesc canonType = type.ConvertToCanonForm(CanonicalFormKind.Specific);
+            if (canonType != type)
+            {
+                // Codegen inlines information about whether the type has a cctor. Since codegen happens
+                // on top of canonical forms, we better agree on whether a type is preinitialized
+                // across all types that share the same canonical form. We should not be preinitializing
+                // types where there's a chance that different substitution produce different answer on this.
```

Comment:
This assume that the pre-initialization interpreter does not ever take different path based on the substitution. It feels like a very fragile assumption.

For example, the static constructor can have a code like `if (typeof(T) == typeof(string))`. We will need be always very careful about not introducing a logic like this in the pre-initializer.

Reply:
I went through all the `GetObject` calls we currently do in the interpreter and reviewed they don't affect the paths code could be taking. Off the top of my head, I only see isinst/castclass and ldtoken in that bucket. I already added a test for the isinst path (in this PR). I can also add ldtoken. We also have this assert. Are there additional mitigations you could think of that would alleviate the concern?